Code/src/main/java/nl/utwente/viskell/ui/components/DisplayBlock.java

```diff
@@ -90,14 +94,21 @@ public void invalidateVisualState() {
         if (inputAnchor.hasConnection()) {
             this.inputType.setText(this.inputAnchor.getStringType());
 
-            try {
-                Optional<GhciSession> ghci = getPane().getGhciSession();
-                if (ghci.isPresent()) {
-                    setOutput(ghci.get().pull(inputAnchor.getFullExpr()));
-                }
-            } catch (HaskellException e) {
-                setOutput("?!?!?!");
-            } 
+            Optional<GhciSession> ghci = getPane().getGhciSession();
+
+            if (ghci.isPresent()) {
+                ListenableFuture<String> result = ghci.get().pull(inputAnchor.getFullExpr());
+
+                Futures.addCallback(result, new FutureCallback<String>() {
+                    public void onSuccess(String s) {
+                        Platform.runLater(() -> setOutput(s));
```

Comment:
The callback gets called on *some* thread (which doesn't have to be, and in this case probably isn't, the JavaFX Application Thread), while the JavaFX Application Thread is the only thread that is allowed to call setOutput. Platform.runLater, of course, schedules work on the JavaFX Application Thread.

A neater solution would be if we could somehow get an `Executor` that corresponds to the JavaFX Application Thread and pass that to `addCallback`, but I haven't been able to find anything like that.